### DISCUSSION TITLE
easier mocking of external APIs with remix

### DISCUSSION THREAD

**Discussion by cellog** [remix-run/remix#4980] [Show and tell / tips]
Hello, I've just released https://github.com/cellog/msw-server-helper#readme as a solution to controlling `msw` mocks from cypress tests.

It's pretty simple: an sqlite database is used as a means of communicating between Cypress and the msw instance inside your remix app. This allows using `cy.enableMockOverride()` inside cypress tests to control which mock to return from an internal API call.

This solves the problem I ran into of "how do I control a mock inside of a remix app?"

Specific examples of use cases I wanted to test but couldn't until now:

- making the API call fail, and verifying the remix app handles it gracefully
- returning a specific result from an API that includes timestamps
- making an API call time out to test that the remix app handles it gracefully

**Comment by cliffordfajardo** [comment]
Thanks for sharing - always great to learn how people are testing their remix apps ✨ 

> a sqlite database is used as a means of communicating between Cypress and the msw instance inside your remix app.
> ...
> This package uses an sqlite database to control which of a list of alternative mock handlers should be used. - [msw-server-helper repo](https://github.com/cellog/msw-server-helper/blob/160f22401f193a079b7abfd80c8c3fb48efad33a/packages/msw-server-helper-cypress/README.md?plain=1#L5)


(1) can you explain a little bit more about how sqlite3 facilitates the communication between cypress & msw? 
(2) Or rephrased, without sqlite3 what challenges were you facing, what couldn't be done before?

**Comment by cellog** [comment]
The problem is that Cypress has no visibility inside the apps it is testing
on the server-side. So if your app makes a request to an external API on
the server (which is recommended practice for remix and nextjs), you can
intercept it with MSW, but what if you want to test 3 different responses?
This allows changing the response mid-test run.
